webknossos-tracingstore/app/com/scalableminds/webknossos/tracingstore/controllers/VolumeTracingController.scala

```diff
@@ -203,6 +219,159 @@ class VolumeTracingController @Inject()(val tracingService: VolumeTracingService
+              ),
+              "dimension_seperator" -> "."
+            ))
+      }
+  }
+
+  def zGroup(token: Option[String], tracingId: String): Action[AnyContent] = Action.async { implicit request =>
+    accessTokenService.validateAccess(UserAccessRequest.readTracing(tracingId), urlOrHeaderToken(token, request)) {
+      Future(Ok(Json.obj("zarr_format" -> 2)))
+    }
+  }
+
+  def rawZarrCube(token: Option[String], tracingId: String, mag: String, cxyz: String): Action[AnyContent] =
+    Action.async { implicit request =>
+      {
+        val combinedToken = urlOrHeaderToken(token, request)
+        accessTokenService.validateAccess(UserAccessRequest.readTracing(tracingId), combinedToken) {
+          for {
+            tracing <- tracingService.find(tracingId) ?~> Messages("tracing.notFound")
+
+            existingMags = tracing.resolutions.map(v => Vec3Int(v.x, v.y, v.z))
+            magParsed <- Vec3Int.fromMagLiteral(mag, allowScalar = true) ?~> Messages("dataLayer.invalidMag", mag)
+            _ <- bool2Fox(existingMags.contains(magParsed)) ?~> Messages("tracing.wrongMag", tracingId, mag) ~> 404
+
+            (c, x, y, z) <- ZarrCoordinatesParser.parseDotCoordinates(cxyz) ?~> Messages("zarr.invalidChunkCoordinates") ~> 404
+            _ <- bool2Fox(c == 0) ~> Messages("zarr.invalidFirstChunkCoord") ~> 404
+            cubeSize = DataLayer.bucketLength
+            request = WebKnossosDataRequest(
+              position = Vec3Int(x * cubeSize * magParsed.x, y * cubeSize * magParsed.y, z * cubeSize * magParsed.z),
+              mag = magParsed,
+              cubeSize = cubeSize,
+              fourBit = Some(false),
+              applyAgglomerate = None,
+              version = None
+            )
+            (data, _) <- tracingService.data(tracingId, tracing, List(request))
+
+            dataWithFallback <- getFallbackLayerDataIfEmpty(tracing, data, mag, cxyz, combinedToken) ?~> "Getting fallback layer failed" ~> 404
+          } yield Ok(dataWithFallback).withHeaders()
+        }
+      }
+    }
+
+  private def getFallbackLayerDataIfEmpty(tracing: VolumeTracing,
+                                          data: Array[Byte],
+                                          mag: String,
+                                          cxyz: String,
+                                          urlToken: Option[String]): Fox[Array[Byte]] = {
+    val isAnnotationEmpty = data.forall(x => x == 0)
+
+    // Since we only do all bucket requests, complete bucket needs to empty so that we use fallback
+    if (isAnnotationEmpty) {
+      val organizationName = tracing.getOrganizationName
+      val dataSetName = tracing.dataSetName
+      val dataLayerName = tracing.getFallbackLayer
+
+      val dataStoreURI = Fox(
+        dataStoreURICache.getOrLoad(
+          (Some(organizationName), dataSetName),
+          K => remoteWebKnossosClient.getDataStoreURIForDataSource(K._1, K._2)
```

Comment:
upper-case single-letter symbols are used for type parameters in scala. please use a normal variable here, maybe `keyTuple` ?